File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/test_descriptor/TestDescriptorUtils.java
package com.teamscale.test_impacted.test_descriptor;

import com.teamscale.client.ClusteredTestDetails;
import com.teamscale.client.TestDetails;
import com.teamscale.test_impacted.commons.IndentingWriter;
import com.teamscale.test_impacted.commons.LoggerUtils;
import com.teamscale.test_impacted.engine.executor.AvailableTests;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestSource;
import org.junit.platform.engine.UniqueId;
import org.junit.platform.engine.UniqueId.Segment;
import org.junit.platform.engine.support.descriptor.ClassSource;
import org.junit.platform.engine.support.descriptor.MethodSource;

import java.util.List;
import java.util.Optional;
import java.util.logging.Logger;
import java.util.stream.Stream;

/** Class containing utility methods for {@link TestDescriptor}s. */
public class TestDescriptorUtils {

	private static final Logger LOGGER = LoggerUtils.getLogger(TestDescriptorUtils.class);

	/** Returns the test descriptor as a formatted string with indented children. */
	public static String getTestDescriptorAsString(TestDescriptor testDescriptor) {
		IndentingWriter writer = new IndentingWriter();
		printTestDescriptor(writer, testDescriptor);
		return writer.toString();
	}

	private static void printTestDescriptor(IndentingWriter writer, TestDescriptor testDescriptor) {
		writer.writeLine(testDescriptor.getUniqueId().toString());
		writer.indent(() -> {
			for (TestDescriptor child : testDescriptor.getChildren()) {
				printTestDescriptor(writer, child);
			}
		});
	}

	/**
	 * Returns true if the {@link TestDescriptor} is an actual representative of a test. A representative of a test is
	 * either a regular test that was not dynamically generated or a test container that dynamically registers multiple
	 * test cases.
	 */
	public static boolean isTestRepresentative(TestDescriptor testDescriptor) {
		boolean isTestTemplateOrTestFactory = isTestTemplateOrTestFactory(testDescriptor);
		boolean isNonParameterizedTest = testDescriptor.isTest() && !isTestTemplateOrTestFactory(
				testDescriptor.getParent().get());
		return isNonParameterizedTest || isTestTemplateOrTestFactory;
	}

	/**
	 * Returns true if a {@link TestDescriptor} represents a test template or a test factory.
	 * <p>
	 * An example of a {@link UniqueId} of the {@link TestDescriptor} is:
	 * <p>
	 * {@code
	 * [engine:junit-jupiter]/[class:com.example.project.JUnit5Test]/[test-template:withValueSource(java.lang.String)]}
	 */
	public static boolean isTestTemplateOrTestFactory(TestDescriptor testDescriptor) {
		if (testDescriptor == null) {
			return false;
		}
		List<Segment> segments = testDescriptor.getUniqueId().getSegments();

		if (segments.isEmpty()) {
			return false;
		}

		String lastSegmentType = segments.get(segments.size() - 1).getType();
		return JUnitJupiterTestDescriptorResolver.TEST_TEMPLATE_SEGMENT_TYPE.equals(
				lastSegmentType) || JUnitJupiterTestDescriptorResolver.TEST_FACTORY_SEGMENT_TYPE.equals(
				lastSegmentType);
	}

	/** Creates a stream of the test representatives contained by the {@link TestDescriptor}. */
	public static Stream<TestDescriptor> streamTestRepresentatives(TestDescriptor testDescriptor) {
		if (isTestRepresentative(testDescriptor)) {
			return Stream.of(testDescriptor);
		}
		return testDescriptor.getChildren().stream().flatMap(TestDescriptorUtils::streamTestRepresentatives);
	}

	/**
	 * Returns the {@link Segment#getValue()} matching the type or {@link Optional#empty()} if no matching segment can
	 * be found.
	 */
	public static Optional<String> getUniqueIdSegment(TestDescriptor testDescriptor, String type) {
		return testDescriptor.getUniqueId().getSegments().stream().filter(segment -> segment.getType().equals(type))
				.findFirst().map(
						Segment::getValue);
	}

	/** Returns {@link TestDetails#sourcePath} for a {@link TestDescriptor}. */
	public static String getSource(TestDescriptor testDescriptor) {
		Optional<TestSource> source = testDescriptor.getSource();
		if (source.isPresent() && source.get() instanceof MethodSource) {
			MethodSource ms = (MethodSource) source.get();
			return ms.getClassName().replace('.', '/');
		}
		if (source.isPresent() && source.get() instanceof ClassSource) {
			ClassSource classSource = (ClassSource) source.get();
			return classSource.getClassName().replace('.', '/');
		}
		return null;
	}

	/** Returns the {@link AvailableTests} contained within the root {@link TestDescriptor}. */
	public static AvailableTests getAvailableTests(TestDescriptor rootTestDescriptor,
												   String partition) {
		AvailableTests availableTests = new AvailableTests();

		TestDescriptorUtils.streamTestRepresentatives(rootTestDescriptor)
				.forEach(testDescriptor -> {
					Optional<String> engineId = testDescriptor.getUniqueId().getEngineId();
					if (!engineId.isPresent()) {
						LOGGER.severe(() -> "Unable to determine engine ID for " + testDescriptor + "!");
						return;
					}

					ITestDescriptorResolver testDescriptorResolver = TestDescriptorResolverRegistry
							.getTestDescriptorResolver(engineId.get());
					Optional<String> clusterId = testDescriptorResolver.getClusterId(testDescriptor);
					Optional<String> uniformPath = testDescriptorResolver.getUniformPath(testDescriptor);
					String source = TestDescriptorUtils.getSource(testDescriptor);

					if (!uniformPath.isPresent()) {
						LOGGER.severe(() -> "Unable to determine uniform path for test descriptor: " + testDescriptor);
						return;
					}

					if (!clusterId.isPresent()) {
						LOGGER.severe(
								() -> "Unable to determine cluster id path for test descriptor: " + testDescriptor);
						return;
					}

					ClusteredTestDetails testDetails = new ClusteredTestDetails(uniformPath.get(), source, null,
							clusterId.get(), partition);
					availableTests.add(testDescriptor.getUniqueId(), testDetails);
				});


		return availableTests;
	}
}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/test_descriptor/TestDescriptorUtils.kt
package com.teamscale.test_impacted.test_descriptor

import com.teamscale.client.ClusteredTestDetails
import com.teamscale.test_impacted.commons.IndentingWriter
import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import com.teamscale.test_impacted.commons.LoggerUtils.getLogger
import com.teamscale.test_impacted.engine.executor.AvailableTests
import org.junit.platform.engine.TestDescriptor
import org.junit.platform.engine.UniqueId
import org.junit.platform.engine.support.descriptor.ClassSource
import org.junit.platform.engine.support.descriptor.MethodSource
import java.util.*
import java.util.stream.Stream

/** Class containing utility methods for [TestDescriptor]s.  */
object TestDescriptorUtils {
	private val LOGGER = createLogger()

	/** Returns the test descriptor as a formatted string with indented children.  */
	@JvmStatic
	fun getTestDescriptorAsString(testDescriptor: TestDescriptor): String {
		val writer = IndentingWriter()
		writer.printTestDescriptor(testDescriptor)
		return writer.toString()
	}

	private fun IndentingWriter.printTestDescriptor(testDescriptor: TestDescriptor) {
		writeLine(testDescriptor.uniqueId.toString())
		indent {
			testDescriptor.children.forEach { child ->
				printTestDescriptor(child)
			}
		}
	}

	/**
	 * Returns true if the [TestDescriptor] is an actual representative of a test. A representative of a test is
	 * either a regular test that was not dynamically generated or a test container that dynamically registers multiple
	 * test cases.
	 */
	@JvmStatic
	fun TestDescriptor.isRepresentative(): Boolean {
		val isTestTemplateOrTestFactory = isTestTemplateOrTestFactory()
		val isNonParameterizedTest = isTest && !parent.get().isTestTemplateOrTestFactory()
		return isNonParameterizedTest || isTestTemplateOrTestFactory
	}

	/**
	 * Returns true if a [TestDescriptor] represents a test template or a test factory.
	 *
	 *
	 * An example of a [UniqueId] of the [TestDescriptor] is:
	 *
	 *
	 * `[engine:junit-jupiter]/[class:com.example.project.JUnit5Test]/[test-template:withValueSource(java.lang.String)]`
	 */
	private fun TestDescriptor.isTestTemplateOrTestFactory(): Boolean {
		val segments = uniqueId.segments

		if (segments.isEmpty()) {
			return false
		}

		val lastSegmentType = segments[segments.size - 1].type
		return JUnitJupiterTestDescriptorResolver.TEST_TEMPLATE_SEGMENT_TYPE == lastSegmentType
				|| JUnitJupiterTestDescriptorResolver.TEST_FACTORY_SEGMENT_TYPE == lastSegmentType
	}

	/** Creates a stream of the test representatives contained by the [TestDescriptor].  */
	private fun TestDescriptor.streamTestRepresentatives(): Stream<TestDescriptor> {
		if (isRepresentative()) {
			return Stream.of(this)
		}
		return children.stream().flatMap {
			it.streamTestRepresentatives()
		}
	}

	/**
	 * Returns the [org.junit.platform.engine.UniqueId.Segment.getValue] matching the type or [Optional.empty] if no matching segment can
	 * be found.
	 */
	fun TestDescriptor.getUniqueIdSegment(type: String): Optional<String> =
		uniqueId.segments.stream()
			.filter { it.type == type }
			.findFirst().map { it.value }

	/** Returns [com.teamscale.client.TestDetails.sourcePath] for a [TestDescriptor].  */
	private fun TestDescriptor.source(): String? {
		val source = source.orElse(null) ?: return null
		return when (source) {
			is MethodSource -> source.className.replace('.', '/')
			is ClassSource -> source.className.replace('.', '/')
			else -> null
		}
	}

	/** Returns the [AvailableTests] contained within the root [TestDescriptor].  */
	@JvmStatic
	fun getAvailableTests(
		rootTestDescriptor: TestDescriptor,
		partition: String
	): AvailableTests {
		val availableTests = AvailableTests()

		rootTestDescriptor.streamTestRepresentatives()
			.forEach { testDescriptor ->
				val engineId = testDescriptor.uniqueId.engineId
				if (!engineId.isPresent) {
					LOGGER.severe { "Unable to determine engine ID for $testDescriptor!" }
					return@forEach
				}

				val testDescriptorResolver = TestDescriptorResolverRegistry.getTestDescriptorResolver(engineId.get())
				val clusterId = testDescriptorResolver!!.getClusterId(testDescriptor)
				val uniformPath = testDescriptorResolver.getUniformPath(testDescriptor)

				if (!uniformPath.isPresent) {
					LOGGER.severe { "Unable to determine uniform path for test descriptor: $testDescriptor" }
					return@forEach
				}

				if (!clusterId.isPresent) {
					LOGGER.severe { "Unable to determine cluster id path for test descriptor: $testDescriptor" }
					return@forEach
				}

				val testDetails = ClusteredTestDetails(
					uniformPath.get(),
					testDescriptor.source(),
					null,
					clusterId.get(),
					partition
				)
				availableTests.add(testDescriptor.uniqueId, testDetails)
			}


		return availableTests
	}
}
